option-bucket/src/lib.rs

```diff
@@ -0,0 +1,380 @@
+        unsafe { self.outer.as_ref().unwrap_unchecked() }
+    }
+
+    /// Converts from `&mut Option<T>::Some` to `&mut T`,
+    /// with the lifetime of this `SomeBucket`.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// # use option_bucket::SomeBucket;
+    ///
+    /// let mut text: Option<String> = Some("Hello, world!".to_string());
+    /// let mut some = SomeBucket::new(&mut text).unwrap();
+    ///
+    /// some.as_mut().push_str(" Happy to be here.");
+    /// assert_eq!(some.as_ref(), "Hello, world! Happy to be here.");
+    /// ```
+    pub fn as_mut(&mut self) -> &mut T {
+        // SAFETY: `outer` is guaranteed to be `Some`
+        // by the invariants of `new_unchecked`
+        unsafe { self.outer.as_mut().unwrap_unchecked() }
+    }
+
+    /// Converts from `&mut Option<T>::Some` to `&mut T`,
+    /// with the lifetime of the original reference,
+    /// consuming this `SomeBucket`.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// # use option_bucket::SomeBucket;
+    ///
+    /// let mut text: Option<String> = Some("Hello, world!".to_string());
+    /// let some = SomeBucket::new(&mut text).unwrap();
+    ///
+    /// some.into_mut().push_str(" Happy to be here.");
+    /// assert_eq!(&text.unwrap(), "Hello, world! Happy to be here.");
+    /// ```
+    ///
+    /// ```compile_fail
+    /// # use option_bucket::SomeBucket;
+    ///
+    /// let mut text: Option<String> = Some("Hello, world!".to_string());
+    /// let some = SomeBucket::new(&mut text).unwrap();
+    ///
+    /// some.into_mut().push_str(" Happy to be here.");
+    /// // can not longer use `some`
+    /// some.as_ref();
+    /// ```
+    pub fn into_mut(self) -> &'a mut T {
+        // SAFETY: `outer` is guaranteed to be `Some`
+        // by the invariants of `new_unchecked`
+        unsafe { self.outer.as_mut().unwrap_unchecked() }
+    }
+
+    /// Sets the value in the `Option<T>::Some`, and returns
+    /// the old value.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// # use option_bucket::SomeBucket;
+    ///
+    /// let mut x = Some(2);
+    /// let mut some = SomeBucket::new(&mut x).unwrap();
+    ///
+    /// let old = some.replace(5);
+    /// assert_eq!(old, 2);
+    /// assert_eq!(x, Some(5));
+    /// ```
+    pub fn replace(&mut self, value: T) -> T {
+        core::mem::replace(self.as_mut(), value)
+    }
+
+    /// Takes the value out of the option, leaving a `None` in its place,
+    /// and consuming this `SomeBucket`.
+    ///
+    /// ```
+    /// # use option_bucket::SomeBucket;
+    ///
+    /// let mut x = Some(vec![1, 2]);
+    /// let some = SomeBucket::new(&mut x).unwrap();
+    ///
+    /// let y = some.take();
+    /// assert_eq!(x, None);
+    /// assert_eq!(y, vec![1, 2]);
+    /// ```
+    pub fn take(self) -> T {
+        // SAFETY: `outer` is guaranteed to be `Some`
+        // by the invariants of `new_unchecked`
+        unsafe { self.outer.take().unwrap_unchecked() }
+    }
+}
+
```

Comment:
When browsing the code I noticed that `NoneBucket` and `SomeBucket` can be redesigned with `#[repr(transparent)]` to accomplish two things:
* Avoid the lifetime parameter.
* Lean more heavily on deriving lifetimes implicitly from `&mut self`.

The declaration and `new_unchecked` constructor of `NoneBucket` would then become this:

```rust
// Note: #[repr(transparent)] ensures that `NoneBucket<T>` is layout transparent with `Option<T>` and that the pointer cast is permitted.
#[repr(transparent)]
pub struct NoneBucket<T> {
    outer: Option<T>,
}

impl<T> NoneBucket<T> {
    pub unsafe fn new_unchecked(opt: &mut Option<T>) -> &mut Self {
        &mut *(opt as *mut _ as *mut Self)
    }
}
```

Note that it's similar (ish) to how [`std::path::Path` works w.r.t. `OsStr`](https://doc.rust-lang.org/src/std/path.rs.html#1927).

Reply:
Good idea, will do.

Reply:
I did notice that this would make `fn take` `unsafe` (since it could no longer take `self` and calling it more than once would be UB). So it might unfortunately not be worth it :/.

Reply:
Ah good point, never mind then.